#!/bin/bash -l
#SBATCH -A uppmax2022-2-30
#SBATCH -M snowy
#SBATCH -p core -n 1
#SBATCH --mem=64gb
#SBATCH -t 24:00:00
#SBATCH -J trimm
#SBATCH --error /error_path/00.FastpLane1.err
#SBATCH --output /output_path/00.FastpLane1.out

##################################################################################################################################
##### This is the command we used for filtering, here you can see the filter for trimming adapters and trimming of poliG > 2 #####
##################################################################################################################################

module load bioinfo-tools
module load fastp
i=/path/to/R1_001.fastq.gz
j=/path/to/R2_001.fastq.gz
h=/path/to/filter/folder
g=MeDIP.R1.nopoliG.fastq.gz #name of the R1 output
f=MeDIP.R2.nopoliG.fastq.gz #name of the R2 output
w=MeDIP.nopoliG.html #name of the report

fastp --thread 4 -z 4 --in1 $i --in2 $j --out1 $h/$g --out2 $h/$f -z 4 --detect_adapter_for_pe --poly_g_min_len 2 -h $h/$w

# The -z option is used to specify the number of worker threads, 
#--in1 $i: Specifies the input file for the first read in a paired-end sequencing experiment. The value of $i is a placeholder for the actual input file name.
#--in2 $j: Specifies the input file for the second read in a paired-end sequencing experiment. The value of $j is a placeholder for the actual input file name.
#--out1 $h/$g: Specifies the output file name for the first read in a paired-end sequencing experiment. The value of $h/$g is a placeholder for the output file name, which will be generated by the program.
#--out2 $h/$f: Specifies the output file name for the second read in a paired-end sequencing experiment. The value of $h/$f is a placeholder for the output file name, which will be generated by the program.
#-z 4: Compress the output file using gzip with a compression level of 4.
#--detect_adapter_for_pe: Automatically detect and trim adapter sequences in paired-end sequencing reads.
#--poly_g_min_len 2: Trims poly-G tails from the end of sequencing reads, with a minimum length of 2 nucleotides.
#-h $h/$w: Specifies the output file for the HTML report generated by the program. The value of $h/$w is a placeholder for the actual output file name. The HTML report summarizes the quality control metrics and data trimming results.
