# GEroNIMO_bioinfo

Hello and welcome to the GEroRIMO workshop on GBS-MeDIP data analysis! 

The objective of this workshop is to guide the attendees in transferring the data from 
i) the facility to the server where the analyses will be performed to 
ii) basic downstream analyses of the identified differentially methylated regions (DMRs).

We will go through basic trimming of the raw data, prepare the reference genome to align the trimmed data and talk about different parameters for these steps.

After that, we will define statistical contrasts, cut-offs, and other basic statistics, according to predefined scientific hypotheses. Finally, we will identify differentially methylated regions (DMRs) among the individuals compared, based on the previously defined contrasts and settings.

Once DMRs are identified, we will introduce you to some useful commands in R to handle DMR data. These commands are employed to: 
i) count the number of CpGs per fragment, 
ii) extract the nucleotide sequence information of the DMR, 
iii) annotate where the DMRs are located in relation to genes within the reference genome of interest (Gallus gallus/Sus Scrofa), and 
iv) generate a final spreadsheet that compiles all the information generated.


On behalf of Carlos, Violeta, Manolis, and me.
Best regards, 


## First ~in silico~ digestion
